[DOCS-7127] Add TMQ info (#932)
Alfresco/docs-alfresco: search-enterprise/latest/config/transactional.md

For multi-valued properties, CMIS QL supports `ANY` semantics from SQL 92. A query against a multi-lingual property, such as title or description, is treated as multi-valued and may match in any language. In the results, you will see the best value for your locale, which may not match the query. Also, ordering will consider any value.

### UPPER() and LOWER()

`UPPER()` and `LOWER()` functions were in early drafts for the CMIS 1.0 specification, but were subsequently dropped. These are not part of the CMIS 1.0 or 1.1 specifications. They are not supported in TMDQ.

### Alfresco FTS QL & TMDQ

It is more difficult to write AFTS queries that use TMDQ as the default behaviour is to use full text queries for text and full text queries cannot be served by the database. Also, special fields like `SITE` and `TAG` that are derived from paths or other nodes will not be handled by the database. `TYPE`, `ASPECT` and the related exact matches will work with TMDQ. All property data types are fine but strings should be less than 1024 characters in length. Text queries have to be prefixed with `=` to avoid full text search.

> Ranges, PATH, and ANCESTOR are not currently supported.

Reviewer: Does Ranges need to be capitalized similar to PATH and ANCESTOR?

Author: Cheers its supposed to be like that

TMDQ can support exact match on all properties (subject to database collation) regardless of the property index configuration in the data model. All text properties can support pattern matching. The database index supports a fixed number of leading characters. The database store a maximum string size before it overflows to another form. Only short form strings can be used in database queries.

Elasticsearch supports exact match on all non-text properties. Text properties only support exact and pattern matches if set to tokenised `both` or `false` in the data model. Elasticsearch provides supports values up to approximately 32,700 UTF-8 bytes.

The following specific CMIS QL fields are supported:

* `cmis:parentId`
* `cmis:objectId`
* `cmis:objectTypeId`
* `cmis:baseTypeId`
* `cmis:contentStreamMimeType`
* `cmis:contentStreamLength`

The following CMIS QL comparison operators are supported:

* `=`, `!=,` `<>`, `<`, `<=`, `>`, `>=`
* `IN`, `NOT IN`, `LIKE`

The following AFTS exact matches and patterns are supported:

* `=<field>:term`
* `=<field>:ter*`
* `=<field>:*erm`

### Support for special fields in TMDQ using AFTS

* PARENT
* TYPE
* ASPECT
* EXACTTYPE
* EXACTASPECT

> > **Note:** CMIS QL does not support any use of CONTAINS() using the database.

Reviewer: Change CONTAINS() to `CONTAINS()`, I think.

Author: Changed cheers

  * Supports ordering for single-valued properties, for example:

    ```sql
    select * from cmis:document where cmis:name <> 'fred' order by cmis:name
    ```

* `integer`, `double`, and `float`
  * Supports comparisons, such as `=`, `<>`, `<`, `<=`, `>=`, `>`, `IN`, `NOT IN`.
  * Supports ordering for single-valued properties.
* `boolean`
  * Support for comparisons `=` and `<>`
  * Supports ordering for single-valued properties.
* `id`
  * Supports `cmis:objectId`, `cmis:baseTypeId`, `cmis:objectTypeId` and `cmis:parentId` fields.
  * Support for comparisons, using `=`, `<>`, `IN` and `NOT IN`.
  * Ordering using a property, which is a CMIS identifier, is not supported.
* `datetime`
  * Supports comparisons, such as `=`, `<>`, `<`, `<=`, `>=`, `>`, `IN` and `NOT IN`.
  * Support ordering for single-valued properties, for example:

    ```sql
    select * from cmis:document where cmis:lastModificationDate = '2010-04-01T12:15:00.000Z' order by
     cmis:creationDate ASC
    ```

> **Note:** While the CMIS URI data type is not supported, multi-valued properties and multi-valued predicates as defined in the CMIS specification are supported. For example,

```sql

> select * from ext:doc where 'test' = ANY ext:multiValuedStringProperty

Author: Ive just removed it as a note. Which I think is fine. Its an interesting one of how to handle situations like that. cheers.

> **Note:** Alfresco Content Services supports TMDQ by default.

## Options supported by Query Languages

Use this information to know what options are supported by the v1 REST API, CMIS Query Language (QL), and Alfresco Full Text Search Query Language (FTS QL).

### v1 REST API and TMDQ

For the v1 REST API, anything that is not a simple query, a filter query, an option that affects these, or an option that affects what is returned for each node in the results, is not supported by TMDQ.

TMDQ supports:

* `query`
* `paging`
* `include`
* `includeRequest`
* `fields`
* `sort`
* `defaults`
* `filterQueries`
* `scope` (single)
* `limits` for permission evaluation

The default limits for permission evaluation will restrict the results returned from TMDQ based on both the number of results processed and time taken. These can be increased, if required.

The v1 REST API does not support TMDQ for:

* `templates`
* `localisation` and `timezone`
* `facetQueries`
* `facetFields`
* `facetIntervals`
* `pivots`
* `stats`

> * `spellcheck`

Reviewer 2: I don't think this is currently supported by the Elasticsearch Connector at all, and I'm not sure about some of the other items in this list. While they are definitely all unsupported for TMDQ then they may also be currently unsupported by ES too, making the paragraph a little misleading.

Author: Removed.

## Transactional metadata queries supported by database

Use this information to understand the queries supported by the database.

The Alfresco Full Text Search (FTS) query text can be used standalone or it can be embedded in CMIS-SQL using the `contains()` predicate function. The CMIS specification supports a subset of Alfresco FTS. For more information on search syntax, see [Alfresco Full Text Search Reference]({% link search-services/latest/using/index.md %}).

**CMIS QL**

The following object types and their sub-types are supported:

* `cmis:document`

    For example:

    ```sql
    select * from cmis:document
    ```

* `cmis:folder`

    For example:

    ```sql
    select * from cmis:folder 
    ```

* Aspects

    For example:

    ```sql
    select * from cm:dublincore 
    ```

## CMIS property data types

The `WHERE` and `ORDER BY` clauses support the following property data types and comparisons:

* `string`
  * Supports comparisons using `=`, `<>`, `<`, `<=`, `>=`, `>`, `IN` and `NOT IN` `LIKE`
  * Supports ordering for single-valued properties

>     For example:

Reviewer 2: This renders on the same line as the previous sentence giving a capital letter in the middle of a sentence.

Author: Right I see, I have adjust how it is coded.

* `d:mltext`: This data type ignores locale. However, if there is more than one locale, the localised values behave as a multi-valued string. Ordering on `mltext` will be undefined as it is effectively multi-valued.

## Configuring transactional metadata query

Configure the transaction metadata query using the subsystem properties.

The common properties used to configure the transactional metadata query for the search subsystems are:

* `query.cmis.queryConsistency`
* `query.fts.queryConsistency`

These properties should be set in the `<TOMCAT_HOME>/shared/classes/alfresco-global.properties` file.

> **Important:** The name of these properties has changed. They were called `solr.query.cmis.queryConsistency` and `solr.query.fts.queryConsistency` but these were deprecated. The old names of the properties will still work with Elasticsearch but it is highly recommended you update them in your configuration.

The default value for these properties is `TRANSACTIONAL_IF_POSSIBLE`. However, you can override it with any of the following permitted values:

* `EVENTUAL`
* `TRANSACTIONAL`

> The `solr.query.cmis.queryConsistency` and `solr.query.fts.queryConsistency` properties can also be set per query on the `SearchParameters` and `QueryOptions` objects in the V1 REST API.

Reviewer 2: ```suggestion
The `query.cmis.queryConsistency` and `query.fts.queryConsistency` properties can also be set per query on the `SearchParameters` and `QueryOptions` objects in the V1 REST API.
```

Use this information to understand the queries supported by the database.

The Alfresco Full Text Search (FTS) query text can be used standalone or it can be embedded in CMIS-SQL using the `contains()` predicate function. The CMIS specification supports a subset of Alfresco FTS. For more information on search syntax, see [Alfresco Full Text Search Reference]({% link search-services/latest/using/index.md %}).

**CMIS QL**

The following object types and their sub-types are supported:

* `cmis:document`

    For example:

    ```sql
    select * from cmis:document
    ```

* `cmis:folder`

    For example:

    ```sql
    select * from cmis:folder 
    ```

* Aspects

    For example:

    ```sql
    select * from cm:dublincore 
    ```

## CMIS property data types

The `WHERE` and `ORDER BY` clauses support the following property data types and comparisons:

* `string`
  * Supports all properties and comparisons, such as `=`, `<>`, `<`, `<=`, `>=`, `>`, `IN`, `NOT IN`, `LIKE`
  * Supports ordering for single-valued properties
    For example:

    ```sql
    select * from cmis:document where cmis:name <> 'fred' order by cmis:name
    ```

* `integer`

>   * Supports all properties and comparisons, such as `=`, `<>`, `<`, `<=`, `>=`, `>`, `IN`, `NOT IN`

Reviewer 2: ```suggestion
  * Supports comparisons, such as `=`, `<>`, `<`, `<=`, `>=`, `>`, `IN`, `NOT IN`
```

Author: rechecked is how you have suggested.

  * Supports ordering for single-valued properties

* `double`
  * Supports all properties and comparisons, such as `=`, `<>`, `<`, `<=`, `>=`, `>`, `IN`, `NOT IN`
  * Supports ordering for single-valued properties

> * `float`

Reviewer 2: I would combine these three since the support is the same for all of them.

Author: Changed.

Author: Is changed correctly rechecked

  * Supports all properties and comparisons, such as `=`, `<>`, `<`, `<=`, `>=`, `>`, `IN`, `NOT IN`
  * Supports ordering for single-valued properties
* `boolean`

>   * Supports properties and comparisons, such as `=` and `<>`

Reviewer 2: ```suggestion
  * Support for comparisons `=` and `<>`
```

Author: rechecked is there as it should be

  * Supports ordering for single-valued properties
* `id`

>   * Supports `cmis:objectId`, `cmis:baseTypeId`, `cmis:objectTypeId`, `cmis:parentId`, `=`, `<>`, `IN`, `NOT IN`

Reviewer 2: ```suggestion
  * Support for `cmis:objectId`, `cmis:baseTypeId`, `cmis:objectTypeId` and `cmis:parentId` fields
  * Support for comparison using `=`, `<>`, `IN`, `NOT IN`
```

Author: Done: search for   * Supports `cmis:objectId`, `cmis:baseTypeId`, `cmis:objectTypeId` and `cmis:parentId` fields
  * Support for comparisons, using `=`, `<>`, `IN`, `NOT IN`

  * Ordering using a property, which is a CMIS identifier, is not supported
* `datetime`

>   * Supports all properties and comparisons `=`, `<>`, `<`, `<=`, `>=`, `>`, `IN`, `NOT IN`

Reviewer 2: ```suggestion
  * Supports comparisons using `=`, `<>`, `<`, `<=`, `>=`, `>`, `IN` and `NOT IN`
```

Author: rechecked is there

  * Supports ordering for single-valued properties
    For example:

    ```sql
    select * from cmis:document where cmis:lastModificationDate = '2010-04-01T12:15:00.000Z' order by
     cmis:creationDate ASC
    ```

> **Note:** While the CMIS URI data type is not supported, multi-valued properties and multi-valued predicates as defined in the CMIS specification are supported. For example,

```sql
select * from ext:doc where 'test' = ANY ext:multiValuedStringProperty
```

## Supported predicates

A predicate specifies a condition that is true or false about a given row or group. The following predicates are supported:

* Comparison predicates, such as `=`, `<>`, `<`, `<=`, `>=`, `>`, `<>`
  * `IN` predicate
  * `LIKE` predicate

    > **Note:** Prefixed expressions perform better and should be used where possible.

* `NULL` predicate 
* Quantified comparison predicate (`= ANY`)

> * Quantified IN predicate (`ANY .... IN (....)`)

Reviewer 2: I've no idea if these two statements work. I'm pretty sure we don't have any tests using `ANY`.

Author: Removed them.

Author: isnt there! Great!

* `IN_FOLDER` predicate function

## Unsupported predicates

> The following predicates are not supported:

Reviewer 2: ```suggestion
The following predicates are not supported by TMDQ:
```

* TEXT search predicate, such as `CONTAINS()` and `SCORE()` 
* `IN_TREE()` predicate

## Supported logical operators

The following logical operators are supported:

* `AND` 
* `NOT`
* `OR`

## Other operators

In the following cases, the query will go to the database but the result might not be as expected. In all other unsupported cases, the database query will fail and fall back to be executed against the Elasticsearch subsystem.

* `IS NOT NULL`
* `IS NULL`: Currently, this operator will only find properties that are explicitly NULL as opposed to the property not existing.
* `SORT`: The multi-valued and `mltext` properties will sort according to one of the values. Ordering is not localized and relies on the database collation. It uses an `INNER JOIN`, which will also filter NULL values from the result set.
* `d:mltext`: This data type ignores locale. However, if there is more than one locale, the localised values behave as a multi-valued string. Ordering on `mltext` will be undefined as it is effectively multi-valued.

> * `UPPER()` and `LOWER()`: Comparison predicates provide additional support for SQL `UPPER()` and LOWER() functions (that were dropped from a draft version of CMIS specification but are supported for backward compatibility).

Reviewer 2: ```suggestion
* `UPPER()` and `LOWER()`: Comparison predicates provide additional support for SQL `UPPER()` and `LOWER()` functions (that were dropped from a draft version of CMIS specification but are supported for backward compatibility).
```

I thought that earlier in the document we wrote that these functions were unsupported.

Author: Removed the text after chat with Martin S

Author: not there I ve checked

## Configuring transactional metadata query

Configure the transaction metadata query using the subsystem properties.

The common properties used to configure the transactional metadata query for the search subsystems are:

* `solr.query.cmis.queryConsistency`
* `solr.query.fts.queryConsistency`

These properties should be set in the `<TOMCAT_HOME>/shared/classes/alfresco-global.properties` file.

The default value for these properties is `TRANSACTIONAL_IF_POSSIBLE`. However, you can override it with any of the following permitted values:

* `EVENTUAL`
* `TRANSACTIONAL`

> The `solr.query.cmis.queryConsistency` and `solr.query.fts.queryConsistency` properties can also be set per query on the `SearchParameters` and `QueryOptions` objects.

Reviewer 2: ```suggestion
The `solr.query.cmis.queryConsistency` and `solr.query.fts.queryConsistency` properties can also be set per query on the `SearchParameters` and `QueryOptions` objects in the Java Public API.
```

Author: The `solr.query.cmis.queryConsistency` and `solr.query.fts.queryConsistency` properties can also be set per query on the `SearchParameters` and `QueryOptions` objects in the V1 REST API.

This is there. I changed the PUBLIC to be V1 as you mentioned somewhere else. If this isnt correct let me know :-)

## Configuring an optional patch for upgrade

> Transactional metadata query requires two optional patches to be applied for full support. If no patch is applied there is no database support.

Reviewer 2: This whole section looks obsolete.

Author: removed

Author: not there

To use or run a query against the `float`, `double`, or `boolean` property data types, you need to run an optional patch that adds the required indexes to the database. To do so, set the following property in the `<TOMCAT_HOME>/shared/classes/alfresco-global.properties` file:

```bash
system.metadata-query-indexes-more.ignored=false 
```

When using all other data types (such as `string`, `integer`, `id`, or `datetime`), to enable the patch that adds the required indexes to the database, set the following property in the `<TOMCAT_HOME>/shared/classes/alfresco-global.properties` file:

```bash
system.metadata-query-indexes.ignored=false 
```

If these optional patches are not run, the metadata query will not be used, regardless of the configuration. This configuration is checked when the subsystem is reloaded.

For a new install, the default behavior is to use the `TRANSACTIONAL_IF_POSSIBLE` metadata queries. For an upgraded system, the `TRANSACTIONAL_IF_POSSIBLE` metadata queries will be used only if the upgrade patches have been run.

> ## Adding optional indexes to a database

Reviewer 2: This section also looks obsolete.

When you are upgrading the database, you can add optional indexes in order to support the metadata query feature. This information lets you know the likely duration of the upgrade and how to do it incrementally.

For large repositories, creating the database indexes to support the transactional metadata query can take some time. To check how long it will take, you can add the first index to the database and note the time taken. The full upgrade is estimated to take less than 10 times this value. However, this can vary depending on the structure of the data, the database, and the size of the repository.

The [SQL patch script](http://dev.alfresco.com/resource/AlfrescoOne/5.0/configuration/alfresco/dbscripts/upgrade/4.2/org.hibernate.dialect.Dialect/metadata-query-indexes.sql) can be run in parts, adding one index at a time. The patch is marked complete by the statement that inserts into alf_applied_patch. The patch can be marked as unapplied using the SQL delete statement.

## Configuring search in Alfresco Share

The following sections describe how to configure search in Alfresco Share.

> ## Controlling permissions checking

Reviewer 2: ```suggestion
### Controlling permissions checking
```

I assume this should be within the section entitled "Configuring search in Alfresco Share"? It would be worth reviewing the number of hashes in other headings below too.

Author: Changed and checked other headings seem ok

Author: its there ive checked

> You can limit the time Alfresco Content Services spends on ensuring that the user executing the search has the necessary permissions to see each result. Setting this limit increases search speed and reduces the use of resources.

Reviewer 2: ```suggestion
TMDQ may take a long time when trying to create a page of results with a sparse result set. You can limit the time Alfresco Content Services spends per TMDQ query by configuring a maximum duration or a maximum number of permission checks before returning. Setting this limit increases search speed and reduces the use of resources but may result in the user receiving a partial page of results for expensive queries.
```

Author: Its there Ive checked

> You can limit both the time spent and the number of documents checked before Alfresco Content Services returns a search query using the `system.acl.maxPermissionCheckTimeMillis` and the `system.acl.maxPermissionChecks` properties. The default values are 10000 and 1000 respectively.

Reviewer 2: ```suggestion
You can limit both the time spent and the number of documents checked before Alfresco Content Services returns search results using the `system.acl.maxPermissionCheckTimeMillis` and the `system.acl.maxPermissionChecks` properties. The default values are 10000 and 1000 respectively.
```

Author: Its there! cool

1. Open the `<classpathRoot>/alfresco-global.properties` file.

2. Set the `system.acl.maxPermissionCheckTimeMillis` property.

    For example, `system.acl.maxPermissionCheckTimeMillis=20000`.

3. Set the `system.acl.maxPermissionChecks` property.

    For example, `system.acl.maxPermissionChecks=2000`.

    > **Note:** If you increase these values and have a query that returns a very large number of results, (a) the search results will take longer to be returned to the user, and (b) the system will spend longer to check permissions, leading to the possibility of performance degradation. If you set these values to a low number, you run the risk of inconsistent search results every time you run the same search. These settings are also applied when paging. So paging the results will only go up to the maximum returned results based on these settings.

> ## Controlling search results

Reviewer 2: ```suggestion
## Limiting search results
```

Author: Its there because I have checked

Use this information to control the maximum number of items that an Alfresco Share search returns.

By default, the Share search feature returns a maximum of 250 search results. You can extend this number of search results to return more than 250 entries.

> 1. Download the [share-config.xml](http://dev.alfresco.com/resource/AlfrescoOne/5.1/configuration/alfresco/share-config.xml) file.

Reviewer 2: Outdated link.

Author: Changed,

Author: File is updated.
I have changed 1. to be

1. Download your `share-config.xml` file found at `tomcat/webapps/share/WEB-INF/classes/alfresco/share-config.xml`, for more see [Share Configuration Extension Point](https://docs.alfresco.com/content-services/latest/develop/share-ext-points/share-config/).

---
title: Transactional metadata query
---

Alfresco Content Services supports the execution of a subset of the CMIS Query Language (CMIS QL) and Alfresco Full Text Search (AFTS) queries directly against the database. Also, the noindex subsystem supports queries only against the database. This collection of features is called transactional metadata query (TMDQ).

TMDQ supports use cases where eventual consistency is not the preferred option.

> The Elasticsearch subsystem is eventually consistent. A change can take any length of time to be reflected in the index, ranging from a few seconds to several minutes. Elasticsearch indexes the metadata and the content of each updated node, in the order in which the nodes were last changed. The rate at which the nodes are indexed is mainly determined by the time it takes to transform the content and the rate at which the nodes are being changed.

Reviewer 2: ```suggestion
The Elasticsearch subsystem is eventually consistent. A change can take any length of time to be reflected in the index, ranging from a few seconds to several minutes. Elasticsearch indexes the metadata and the content of each updated node, in the order in which the nodes were last changed. The indexing components will try to index information about nodes as fast as possible, but content indexing is likely to be limited by the time needed to extract text from the files and all indexing will be affected by the rate at which the nodes are being changed.
```

Author: Rechecked and is there.

> The Elasticsearch subsystem is eventually consistent. A change can take any length of time to be reflected in the index, ranging from a few seconds to several minutes. Elasticsearch indexes the metadata and the content of each updated node, in the order in which the nodes were last changed. The indexing components will try to index information about nodes as fast as possible, but content indexing is likely to be limited by the time needed to extract text from the files and all indexing will be affected by the rate at which the nodes are being changed.

Reviewer 3: We have made effort to improve eventual consistency with Elasticsearch, and I think we should update this para to reflect this. Suggest: 
Replace:
"A change can take any length of time to be reflected in the index, ranging from a few seconds to several minutes."
With
"The amount of time a change a change takes to reflect in the index is normally less than 1 second, but can be longer under heavy load, or for complex/cascading updates."

Author: Re checked wasnt there. Sorry.

> Some queries can be executed both transactionally against the database or with eventual consistency against the Solr index. Only queries using the AFTS or CMIS query languages can be executed against the database. The Lucene query language cannot be used against the database whereas, `selectNodes` (XPATH) on the Java API always goes against the database, walking and fetching nodes as required.

Reviewer 2: ```suggestion
Some queries can be executed both transactionally against the database or with eventual consistency against the Elasticsearch index. Only queries using the AFTS or CMIS query languages can be executed against the database. The Lucene query language cannot be used against the database whereas, `selectNodes` (XPATH) on the Java API always goes against the database, walking and fetching nodes as required.
```

Author: Rechecked and Is there.,

> Improvements to tracking results in less lag to metadata indexing. Metadata updates are impacted less by content indexing or the bulk updates to PATH for `move`, `rename`, `link` and, `unlink` operations.

Reviewer 2: I don't understand what this paragraph is getting at - perhaps it's in reference to a particular release? I'd suggest removing it completely.

Reviewer 3: Agree. Elasticsearch docs should not have any references to tracking.

Author: Removed.

Author: Tracking is not in the docs rechecked.

Author: It is still not there.

> The database can only be used for a subset of all the queries. These queries can be in the CMIS QL or AFTS QL. CMIS QL expressions are more likely to use TMDQ because of the default behavior to do exact matches. AFTS QL defaults to full text search and uses constructs not supported by the database engine. For example, PATH queries.

Reviewer 2: This paragraph has a large overlap with the paragraph starting "Some queries can be executed both transactionally..."

Author: I removed the second one as they were very similar.

Author: Re checked the second one isnt there,

In general, TMDQ does not support:

> * Structural queries, full text search, and special fields: This includes SITE that are derived from structure and long strings (> 1024 characters). Text fields support exact and pattern-based matching subject to the database collation. Filter queries are rewritten along with the main query to create one large query. Ordering is fine, but again subject to database collation for text.

Reviewer 2: This bullet point seems confused. I think it could be three bullet points:
* Structural queries (I'm not exactly sure which structures are/aren't supported)
* Full text search
* Special fields (like SITE)

In addition it looks like there's some information about the impact of database collation that should be extracted from this list of bullet points and put in a separate note.

Author: Done. Check how it looks.

Author: Rechecked and is there.

* Faceting.
* Any aggregation: This includes counting the total number of matches for the query.

> AFTS and CMIS queries are parsed to an abstract form. This is then sent to an execution engine. There are two execution engines: the database and the Elasticsearch index. The default is to try the database first and fall back to the Solr index, if the query is not supported against the database. This is configurable for a search subsystem and per query using the Java API.

Reviewer 3: Reference to Solr needs replacing

Author: Replaced.

To support TMDQ:

> * Alfresco Content Services supports TMDQ by default.

Reviewer 2: Maybe get rid of "To support TMDQ" and leave this as a note without a bullet point?
```
Alfresco Content Services has TMDQ support turned on by default.
```

Author: Done.

Author: Rechecked is there
> **Note:** Alfresco Content Services supports TMDQ by default.

> ## Features

Reviewer 2: This whole section seems to be basically the same information as above.

Author: I have removed it let me know if you have another suggestions.

Author: Rechecked its not there.

* perform transactional execution of queries;
* execute queries transactionally, when possible, and fall back to eventual consistency; or
* always execute eventual consistency.
* When queries are executed against the database:
* Hidden nodes will be returned by the database, as they are in Alfresco Content Services 5.0.
* Large result sets are not supported because Alfresco Content Services does not evaluate permissions in query but as a post filter.
* Counts will not reflect the number of nodes that match the query.
* The `SearchParameters` and `QueryOptions` objects can be used to override this behaviour per query.
Alfresco Content Services supports the execution of a subset of the CMIS Query Language (CMIS QL) and Alfresco Full Text Search (AFTS) queries directly against the database. Also, the noindex subsystem supports queries only against the database. This collection of features is called transactional metadata query (TMDQ).

## Options supported by Query Languages

Use this information to know what options are supported by the Public API, CMIS Query Language (QL), and Alfresco Full Text Search Query Language (FTS QL).

### Public API and TMDQ

> From public API, anything that is not a simple query, a filter query, an option that affects these, or an option that affects what is returned for each node in the results, is not supported by TMDQ.

Reviewer 2: Which public API? Is this talking about the v1 REST API?

Author: I have changed the references to be V1 Rest Api. What else could it be. Thanks.

Author: I changed all references from Public API to be V1 REST API.
No Public in the docs rechecked.

TMDQ supports:

* `query`
* `paging`
* `include`
* `includeRequest`
* `fields`
* `sort`
* `defaults`
* `filterQueries`
* `scope` (single)
* `limits` for permission evaluation

The default limits for permission evaluation will restrict the results returned from TMDQ based on both the number of results processed and time taken. These can be increased, if required.

The Public API does not support TMDQ for:

* `templates`
* `localisation` and `timezone`
* `facetQueries`
* `facetFields`
* `facetIntervals`
* `pivots`
* `stats`
* `spellcheck`
* `highlight`
* `ranges facets`

> Some of these will be ignored and produce transactional results; others will fail and be eventual.

Reviewer 2: ```suggestion
The use of these with TMDQ is undefined. Some of these options will be ignored and results will come from the database; others will cause the database query to fail and ACS will fail over to return results from the search index.
```

> The Public API ignores the SQL select part of a CMIS query and generate the results as it would do for AFTS.

Reviewer 2: ```suggestion
The Public API ignores the SQL `SELECT` part of a CMIS query and generates the results as it would do for AFTS.
```

Author: added this in manually as Github didnt do it.

Author: The V1 REST API ignores the SQL `SELECT` part of a CMIS query and generates the results as it would do for AFTS.

Rechecked is there.

### CMIS QL & TMDQ

> For CMIS QL, all expressions except for `CONTAINS()`, `SCORE()`, and `IN_TREE()` can now be executed against the database. Most data types are supported except for the CMIS uri and html types. Strings are supported but only if there are 1024 characters or less in length.

Reviewer 2: ```suggestion
For CMIS QL, all expressions except for `CONTAINS()`, `SCORE()`, and `IN_TREE()` can now be executed against the database. Most data types are supported except for the CMIS uri and html types. Strings are supported but only if they are 1024 characters or less in length.
```

Author: Rechecked new text is there.

> In Alfresco One 5.0, `OR`, decimal, and boolean types were not supported; it is only from Alfresco One 5.1 onwards that they are supported.

Author: Removed.

> In Alfresco One 5.0, `OR`, decimal, and boolean types were not supported; it is only from Alfresco One 5.1 onwards that they are supported.

Author: Its still not there !

Primary and secondary types are supported and require inner joins to link them together. You can skip joins to secondary types from the fetch in CMIS using the public API. You would need an explicit `SELECT` list and supporting joins from a CMIS client. You still need joins to secondary types for predicates and ordering. As CMIS SQL supports ordering as part of the query language, you have to do it there and not via the Public API sort.

For multi-valued properties, CMIS QL supports `ANY` semantics from SQL 92. A query against a multi-lingual property, such as title or description, is treated as multi-valued and may match in any language. In the results, you will see the best value for your locale, which may not match the query. Also, ordering will consider any value.

### UPPER() and LOWER()

`UPPER()` and `LOWER()` functions were in early drafts for the CMIS 1.0 specification, but were subsequently dropped. These are not part of the CMIS 1.0 or 1.1 specifications. They are not supported in TMDQ.

### Alfresco FTS QL & TMDQ

> It is more difficult to write AFTS queries that use TMDQ as the default behaviour is to use full text queries for text. These cannot go against the database. Also, special fields like `SITE` and `TAG` that are derived from the structure will not go to the database. `TYPE`, `ASPECT` and the related exact matches work fine with TMDQ. All property data types are fine but strings should be less than 1024 characters in length. Text queries have to be prefixed with `=` to avoid full text search. Additionally, `PARENT` is supported but `OR` is supported from Alfresco One 5.1 onwards.

Reviewer 2: ```suggestion
It is more difficult to write AFTS queries that use TMDQ as the default behaviour is to use full text queries for text and full text queries cannot be served by the database. Also, special fields like `SITE` and `TAG` that are derived from paths or other nodes will not be handled by the database. `TYPE`, `ASPECT` and the related exact matches will work with TMDQ. All property data types are fine but strings should be less than 1024 characters in length. Text queries have to be prefixed with `=` to avoid full text search.
```

Ranges, PATH, and ANCESTOR are not currently supported.

### Database & TMDQ

Some differences between the database and TMDQ:

* The database has specific fixed collation as defined by the database schema. This affects all string comparisons, such as ordering or case sensitivity in equality. Elasticsearch uses Java localised collation and supports more advanced ordering and multi-lingual fields. The two engines can produce different results for lexical comparison, case sensitivity, ordering, or when using `mltext` properties.

> * The database results include hidden nodes. You can exclude them in the query. The Elasticsearch index results will never include hidden nodes and respects the index control aspect.

Reviewer 2: I wonder if this is still true with Elasticsearch. We include lots of nodes in the index because they are needed for path indexing.

Author: Cheers @tpage-alfresco which bullet point are you asking me to remove?

Author: Removed 

* The database results include hidden nodes. You can exclude them in the query. The Elasticsearch index results will never include hidden nodes and respects the index control aspect.

Author: still not there

> * The database post filters the results to apply permissions. As a result, no total count can be provided and large result sets are not well supported. This also affects paging behaviour. Permission evaluation is truncated by time or number of evaluations. TMDQ is not intended to scale to more than 10s of thousands of nodes. It will not perform well for users who can read one node in a million. It cannot and will not tell you how many results matched the query. To do this could require an inordinate number of permission checks. It does enough to give you the page requested. The Elasticsearch index can apply permissions at query and facet time to billions of nodes. For the same reason, do not expect any aggregation support in TMDQ.

Reviewer 2: ```suggestion
* The database post filters the results to apply permissions. As a result, no total count can be provided and large result sets are not well supported. This also affects paging behaviour. Permission evaluation is truncated by time or number of evaluations. TMDQ is not intended to scale to tens of thousands of nodes. It will not perform well for users who can only read one node in a million. It cannot tell you how many results matched the query and cannot support aggregations. It will try to do enough to give you the page requested. The Elasticsearch index can apply permissions at query and facet time, allowing queries to scale to billions of nodes.
```

> * `CONTAINS()` support is complicated. The pure CMIS part of the query and `CONTAINS()` part are melded together into a single abstract query representation. By default, in CMIS the `CONTAINS()` expression implies full text search, so the queries will go to the Elasticsearch index.

Reviewer 2: I don't understand why this is complicated. It sounds like TMDQ does not support `CONTAINS()` because this check runs against the text content of nodes.

Author: I adjusted the sentence slightly and took out the complicated bit. Not a word we use in the docs. Let me know if you are happy with that.

Author: This wast changed, its changed now. Sorry

* The database does not score. It will return results in some order that depends on the query plan, unless you ask for specific ordering. A three part `OR` query, where some documents match more than one constraint, is treated as equal. For Elasticsearch index queries, the more parts of an `OR` match, the higher is the score. The docs that match more optional parts of the query will come higher up.

> * Queries from Share will not use TMDQ as they will most likely have a full text part to the query and ask for facets.

Reviewer 2: Additionally we should include:
```
* Exact term search will behave differently when executed against the database or the search index. This is due to how tokenisation is applied to strings in the search index.
```

See https://hub.alfresco.com/t5/alfresco-content-services-blog/exact-term-queries-in-search-services-2-0/ba-p/302200 for more details (SS 1.4 acted more similarly to TMDQ than SS 2.0 and SE 3.0).

Reviewer 2: I see the next section is all about exact matches, so maybe that already covers this, but it seems odd to miss it off the bullet point list here.

Author: Added bullet and link.

Author: Still there rechecked

### Exact match and patterns

TMDQ can support exact match on all properties (subject to database collation) regardless of the property index configuration in the data model. All text properties can support pattern matching. The database index supports a fixed number of leading characters. The database store a maximum string size before it overflows to another form. Only short form strings can be used in database queries.

Elasticsearch supports exact match on all non-text properties. Text properties only support exact and pattern matches if set to tokenised `both` or `false` in the data model. Elasticsearch provides supports values up to approximately 32,700 UTF-8 bytes.

The following specific CMIS QL fields are supported:

* `cmis:parentId`

> * `cmis:objectcId`

Reviewer 2: ```suggestion
* `cmis:objectId`
```

Author: Has been updated correctly

* `cmis:objectTypeId`
* `cmis:baseTypeId`
* `cmis:contentStreamMimeType`
* `cmis:contentStreamLength`

The following CMIS QL comparison operators are supported:

* `=`, `!=,` `<>`, `<`, `<=`, `>`, `>=`
* `IN`, `NOT IN`, `LIKE`

The following AFTS exact matches and patterns are supported:

* `=<field>:term`
* `=<field>:ter*`
* `=<field>:*erm`

> ### Full text search for CMIS QL and AFTS

Reviewer 2: I'm not sure I understand what this section means. It doesn't mention TMDQ and the contents of the section don't mention full text search (apart from the acronym AFTS).

Author: Ive removed it,
